cmd/segment-reaper/detect_test.go

```diff
@@ -0,0 +1,635 @@
+						expectedObj.segments = math.MaxUint64 >> (int(maxNumOfSegments) - numSegments + 1)
+					}
+
+					expectedNumSegments += numSegments
+					if inline {
+						expectedInlineSegments++
+						expectedRemoteSegments += (numSegments - 1)
+					} else {
+						expectedRemoteSegments += numSegments
+					}
+				}
+			}
+
+			ctx := testcontext.New(t)
+			defer ctx.Cleanup()
+
+			for _, objSeg := range objSegments {
+				err := obsvr.processSegment(ctx.Context, objSeg.path, objSeg.pointer)
+				require.NoError(t, err)
+			}
+
+			assert.Equal(t, projID.String(), obsvr.lastProjectID, "lastProjectID")
+			assert.Equal(t, expectedInlineSegments, obsvr.inlineSegments, "inlineSegments")
+			// newObject if returns an inline segment is always the last
+			assert.Equal(t, expectedInlineSegments, obsvr.lastInlineSegments, "lastInlineSegments")
+			assert.Equal(t, expectedRemoteSegments, obsvr.remoteSegments, "remoteSegments")
+
+			if assert.Equal(t, len(expectedObjects), len(obsvr.objects), "objects number") {
+				for cluster, bucketObjs := range obsvr.objects {
+					expBucketObjs, ok := expectedObjects[cluster.bucket]
+					if !ok {
+						t.Errorf("bucket '%s' shouldn't exist in objects map", cluster.bucket)
+						continue
+					}
+
+					if !assert.Equalf(t, len(expBucketObjs), len(bucketObjs), "objects per bucket (%s) number", cluster.bucket) {
+						continue
+					}
+
+					for expPath, expObj := range expBucketObjs {
+						if !assert.Contains(t, bucketObjs, expPath, "path in bucket objects map") {
+							continue
+						}
+
+						obj := bucketObjs[expPath]
+						assert.Equal(t, expObj.expectedNumberOfSegments, obj.expectedNumberOfSegments, "Object.expectedNumSegments")
+						assert.Equal(t, expObj.hasLastSegment, obj.hasLastSegment, "Object.hasLastSegment")
+
+						assert.Equal(t, expObj.skip, obj.skip, "Object.skip")
+						if !expObj.skip {
+							assert.Equal(t, expObj.segments, obj.segments, "Object.segments")
+						}
+					}
+				}
+			}
+		})
+	})
+
+	t.Run("analyzeProject", func(t *testing.T) {
+		t.Run("object without last segment", func(t *testing.T) {
+			var objectsMap ObjectsMap
+			{ // Generate an objects without last segment
+				const (
+					bucketName = "analyzeBucket"
+					objPath    = storj.Path("analyzePath")
+				)
+				var segments uint64
+				{
+					numSegments := rand.Intn(62) + 1
+					segments = math.MaxUint64 >> numSegments
+				}
+
+				projID, err := uuid.New()
+				require.NoError(t, err)
+				objectsMap = ObjectsMap{
+					Cluster{
+						projectID: projID.String(),
+						bucket:    bucketName,
+					}: map[storj.Path]*Object{
+						objPath: {
+							segments:       segments,
+							hasLastSegment: false,
+						},
+					},
+				}
+			}
+
+			var (
+				buf   = &bytes.Buffer{}
+				obsvr = Observer{
+					db:      teststore.New(),
+					writer:  csv.NewWriter(buf),
+					objects: objectsMap,
+				}
+			)
+
+			ctx := testcontext.New(t)
```

Comment:
nit: usually we are creating `testcontext.New(t)` at the beginning of the test. I like this convention but this is nit for me.

Reply:
I've created it here because it's close to where it's used.
I don't mind to change it, but I'd like to hear that the second reviewer prefers to move to the top too.

Reply:
@kaloyan-raev express your opinion on this please

Reply:
I don't really care where it is.

Reply:
@mniewrzal if you have a strong feeling about that the code will be easy to understand by Storj folks, let me know and I do the change.

Reply:
I marked it as nit so don't care too much :) may stay like that

Reply:
In the end, this has been moved, because a requested change has made the need to use the context earlier in the function.